src/playlist-maker.js

```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
+      return -1;
+    }
 
-      let nextIndex;
+    // When repeating, loop back to the end of the playlist.
+    if (playlist.repeat_ && current === 0) {
+      return playlist.indexOfLast();
+    }
 
-      // Repeat
-      if (playlist.repeat_) {
-        nextIndex = playlist.currentIndex_ + 1;
-        if (nextIndex > list.length - 1) {
-          nextIndex = 0;
-        }
+    // Don't go past the beginning of the playlist.
+    return Math.max(current - 1, 0);
+  };
 
-      // Don't go past the end of the playlist.
-      } else {
-        nextIndex = Math.min(playlist.currentIndex_ + 1, list.length - 1);
-      }
+  /**
+   * Plays the first item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if the list is empty.
+   */
+  playlist.first = () => {
+    if (list.length) {
+      return list[playlist.currentItem(0)];
+    }
 
-      // Make the change
-      if (nextIndex !== playlist.currentIndex_) {
-        return list[playlist.currentItem(nextIndex)];
-      }
-    },
+    playlist.currentIndex_ = -1;
+  };
+
+  /**
+   * Plays the last item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if the list is empty.
+   */
+  playlist.last = () => {
+    if (list.length) {
+      return list[playlist.currentItem(playlist.indexOfLast())];
+    }
+
+    playlist.currentIndex_ = -1;
+  };
+
+  /**
+   * Plays the next item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if on last item.
+   */
+  playlist.next = () => {
+    const index = playlist.indexOfNext();
+
+    if (index !== playlist.currentIndex_) {
+      return list[playlist.currentItem(index)];
+    }
+  };
+
+  /**
+   * Plays the previous item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if on first item.
+   */
+  playlist.previous = () => {
+    const index = playlist.indexOfPrevious();
+
+    if (index !== playlist.currentIndex_) {
+      return list[playlist.currentItem(index)];
+    }
+  };
+
+  /**
+   * Get/set auto-advance on the playlist.
+   *
+   * @param  {number} [delay]
+   *         The number of seconds to wait before each auto-advance.
+   *         Leave this argument off to get the current auto-advance delay.
+   *
+   * @return {number|null}
+   *         The current delay.
+   */
+  playlist.autoadvance = (delay) => {
+    if (delay !== undefined) {
+      autoadvance.setup(playlist.player_, delay);
+    }
+  };
+
+  /**
+   * Sets `repeat` option, which makes the "next" video of the last video in
+   * the playlist be the first video in the playlist.
+   *
+   * @param  {boolean} [val]
+   *         The value to set repeat to
+   *
+   * @return {boolean}
+   *         The current value of repeat
+   */
+  playlist.repeat = (val) => {
+    if (val !== undefined) {
+      playlist.repeat_ = !!val;
```

Comment:
Should we return here so that getter/setter functionality isn't mixed? I think that's what we do for video.js